### ISSUE TITLE
Upgrading cluster from 1.7 to 2.3 mapper attachment plugin copied data but not mapping settings

### ISSUE THREAD

**Issue by pholly** [elastic/elasticsearch#18928]
**Elasticsearch version**:
Upgrading from 1.7 to 2.3.3

**Mapper Attachment version**
Upgrading from 2.7.1(version targeted for es1.7) to 2.3.3

**JVM version**:
1.8.0_91

**OS version**:
Windows Server 2012 R2 Standard

**Description of the problem including expected versus actual behavior**:
I'm upgrading our 1.7 cluster to the latest 2.3 cluster. We have 2 nodes but the topology of our cluster isn't relevant for this issue.

After successfully upgrading our cluster I noticed the mapping for our attachment fields did not carry over. Starting with mapper attachment version 3.0.3 the meta field containing the indexed attachment content changed from the name of the field to "content". The upgrade successfully copied my data to the new "content" meta field but the field settings (analyzer, term vectors) were not copied over. This is important because the data was indexed using English analyzer and that's how it's stored in the field. I verified this by returning script field document("field.content") to see what data is actually in the field.

This concrete example will explain it much better:

This is my attachment field mapping in ES1.7
```javascript
"contentPdf": {
        "path": "full",
        "type": "attachment",
        "fields": {
          "date": {
            "format": "dateOptionalTime",
            "type": "date"
          },
          "keywords": {
            "type": "string"
          },
          "content_type": {
            "type": "string"
          },
          "author": {
            "type": "string"
          },
          "name": {
            "type": "string"
          },
          "language": {
            "type": "string"
          },
          "title": {
            "type": "string"
          },
          "contentPdf": {
            "analyzer": "english",
            "term_vector": "with_positions_offsets",
            "store": true,
            "type": "string",
            "fields": {
              "standard": {
                "analyzer": "standard",
                "type": "string"
              }
            }
          },
          "content_length": {
            "store": true,
            "type": "string"
          }
        }
```
There you can see the meta field with the actual indexed content is called `contentPdf` - has to be the same name as the property field name. There I set an `english` analyzer, `term_vector`, `store`, and a sub field with a `standard` analyzer.

After upgrading both ES and the mapper plugin to 2.3.3 this is my mapping
```javascript
"contentPdf": {
        "path": "full",
        "type": "attachment",
        "fields": {
          "date": {
            "format": "epoch_millis||dateOptionalTime",
            "type": "date"
          },
          "keywords": {
            "type": "string"
          },
          "content_type": {
            "type": "string"
          },
          "contentPdf.content": {
            "type": "string",
            "index_name": "contentPdf"
          },
          "author": {
            "type": "string"
          },
          "name": {
            "type": "string"
          },
          "language": {
            "type": "string"
          },
          "title": {
            "type": "string"
          },
          "content_length": {
            "store": true,
            "type": "string"
          }
        }
```
As you can see, the field properties were not copied over from `contentPdf` meta field to `contentPdf.content`. The plugin also added `index_name` for backwards compatibility I believe.

I verified the data was copied by running a query and specifying script fields so I could see what was actually in the index:
```javascript
"script_fields": {
    "pdf content": {
      "script": "doc['contentPdf']"
    }
  },
```
**Problems this causes**
1. Attachment field can't be queried properly because the analyzer info is gone. Since I was using an `english` analyzer all words were indexed with their stemmed form but stemming is not applied when queried. Future documents will be indexed using default analyzer (`standard`) for the content field.
2. Hit highlighting will not work without `term_vectors: with_positions_offsets` and `store: true`.

**Solution if this is not fixed**
After I realized the mapping was not copied over, I reverted to ES1.7 and attempted to add a `content` meta field with correct settings to my `contentPdf` property using PUT mapping API, hoping that during the upgrade the data will be copied to a field that already has the correct mapping. Unfortunately attachment fields do not let you add any other meta fields. I verified this by creating another attachment field with test meta fields and my test meta fields were not taken. 

The only alternative I see is to reindex into a 2.3 cluster then switch over to the new cluster. This is what we are preparing to do right now.

**Steps to reproduce**:
 1. Use PUT mapping API to create a type with an `attachment` property and set `"term_vector": "with_positions_offsets",
            "store": true` on the meta field with the same name as the property. E.g.
```javascript
{
  "TestDoc": {
     "properties" :{
        "myAttachment": {
        "type": "attachment",
        "fields": {
          "myAttachment": {
            "term_vector": "with_positions_offsets",
            "store": true,
            "type": "string"
          }
     }
 }
}
```
 2. Index a document with a base64 encoded binary file as the content of the `myAttachment` field
 3. Shut down all nodes in cluster, upgrade ES and mapper plugin to 2.3.3, restart cluster
4.  Data will now be in `myAttachment.content` field but the mappings for that field will be missing.

Thanks!

**Comment by clintongormley** [comment]
Hi @pholly 

I'm afraid this is an upgrade bug we've just recently discovered and you are correct: your only solution is to reindex these documents.  To compound matters, `contentPdf.content` can't have sub-fields, and doesn't support `copy_to` either.  I realise that this is a real problem, but I can't offer you much of a solution in 2.x.

In 5.x, we have deprecated the mapper-attachements plugin in favour of the ingest-attachment plugin.  This will extract your document during ingestion so that fields like content, author, etc will actually become part of the `_source` and so will work like any other non-magical field.

Sorry I can't give you better news.
